cmake/multi_image.cmake

```diff
@@ -155,8 +156,7 @@ function(add_child_image_from_source name sourcedir)
     BINARY_DIR ${CMAKE_BINARY_DIR}/${name}
     BUILD_BYPRODUCTS ${${name}_BUILD_BYPRODUCTS} # Set by shared_vars.cmake
     CONFIGURE_COMMAND ""
-    BUILD_COMMAND ${CMAKE_MAKE_PROGRAM}
-    ${EXTRA_MULTI_IMAGE_BUILD_OPT} # E.g. -v
+    BUILD_COMMAND ${CMAKE_COMMAND} --build .
```

Comment:
I don't see the need to change this, but I guess it doesn't hurt.

Reply:
For the build part there is normally no difference between invoking `cmake --build` or the build command directly.

But `cmake --build` allows cmake to add additional flags to build command based upon the CMake configuration / generation, in case those were not added by the generator.

It also aligns this code to not have to consider the build command directly.

Reply:
> But cmake --build allows cmake to add additional flags to build command based upon the CMake configuration / generation, in case those were not added by the generator.

This is not clear to me, could you give an example?

Reply:
an additional reason that shows the difference between: `${CMAKE_MAKE_PROGRAM}` and `${CMAKE_COMMAND} --build .` is in the example in https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1650#discussion_r359816575

`${CMAKE_MAKE_PROGRAM}` will always use the parent make program, whereas `${CMAKE_COMMAND} --build .` actually allow the user to specify make program for sub-images, such as `-Dspm_CMAKE_MAKE_PROGRAM=ninja-explain` :wink:

Reply:
Does -Dspm_CMAKE_MAKE_PROGRAM=ninja-explain not work also?

Reply:
> > But cmake --build allows cmake to add additional flags to build command based upon the CMake configuration / generation, in case those were not added by the generator.
> 
> This is not clear to me, could you give an example?

Yes, one example is control of number of parrallel jobs:

https://cmake.org/cmake/help/latest/envvar/CMAKE_BUILD_PARALLEL_LEVEL.html
Which means `cmake --build` passes this info to the sub-image build command, but `${CMAKE_MAKE_PROGRAM}` would not
as example, calling `ninja` on top level but setting `CMAKE_BUILD_PARALLEL_LEVEL-1` ensure sub-images are built in single parallel, which you can judge based on time:
```
$ time CMAKE_BUILD_PARALLEL_LEVEL=1 ninja
real    0m19.690s
user    0m24.619s
sys     0m5.440s
```

```
$ time ninja 
real    0m13.645s
user    0m28.804s
sys     0m6.611s
```
In this example, it is the spm child image that takes longer time to build.

Note: you can also use `cmake --build .` on top level if you want, then both parent and child images are built with parallel=1

Reply:
This is also possible with "ninja -j", I don't see how "cmake --build" opens up opportunities, I see it only limiting what we can do.

Reply:
cause in that case you would need to add `ninja -j` to the sub-image build, hence reinvoke `cmake` to set / clear the flag.

With `CMAKE_BUILD_PARALLEL_LEVEL=1 ninja` you can decide to add this on the fly, without re-invoking `cmake`, because `cmake --build` is used in external project instead of `${CMAKE_MAKE_PROGRAM}`

Reply:
>  I see it only limiting what we can do.

no, nothing is being limited by using `cmake --build .`, quite the opposite.
`cmake --build` supports `--`, so you can do:
`cmake --build . -- -v`
https://cmake.org/cmake/help/latest/manual/cmake.1.html#build-a-project
```
--
    Pass remaining options to the native tool.
```

but on top of that, we get additional possibilities built into CMake.

Reply:
> With CMAKE_BUILD_PARALLEL_LEVEL=1 ninja you can decide to add this on the fly, without re-invoking cmake, because cmake --build is used in external project instead of ${CMAKE_MAKE_PROGRAM}

That is nice, but how will I be able to do the ninja-explain trick when we no longer use CMAKE_MAKE_PROGRAM to construct the external build command?

Perhaps we could have an overridable MULTI_IMAGE_BUILD_COMMMAND that defaulted to "cmake --build ." ?

Reply:
> cmake --build supports --, so you can do:

I was not aware, could we have a secret MULTI_IMAGE_BUILD_OPT to populate it so I can debug then?

Reply:
An additional example:
Run CMake as normal:
```
$ cmake -GNinja -DBOARD=nrf9160_pca10090ns ..
```

Normal build
```
$ ninja
```

Verbose output on just sub-images:
```
$ VERBOSE=True ninja
```
Verbose output on parent only:
```
$ ninja -v
```
Verbose output on parent and child images:
```
$ VERBOSE=True cmake --build .
```
or
```
$ VERBOSE=True ninja -v
```

All of this without re-running CMake :wink:

Reply:
I'm no longer against using cmake --build, I was not aware of it's ability pass arbitrary arguments. But I am against the need for the ninja-explain trick (due to issues on Windows), and would like to have secret MULTI_IMAGE_BUILD_OPT variable for debug reasons.

Reply:
> > With CMAKE_BUILD_PARALLEL_LEVEL=1 ninja you can decide to add this on the fly, without re-invoking cmake, because cmake --build is used in external project instead of ${CMAKE_MAKE_PROGRAM}
> 
> That is nice, but how will I be able to do the ninja-explain trick when we no longer use CMAKE_MAKE_PROGRAM to construct the external build command?
> 

Because if you specify `CMAKE_MAKE_PROGRAM=<myProgram>`, that will be placed in the CMakeCache.txt, and then used by `cmake --build .`
So if you do: 
`cmake -DCMAKE_MAKE_PROGRAM=MyNinjaA -Dspm_CMAKE_MAKE_PROGRAM=MyNinjaB ..` then 
`<parent>::CMakeCache.txt` contains:
```
//No help, variable specified on the command line.
CMAKE_MAKE_PROGRAM:UNINITIALIZED=MyNinjaA
```
and 
`spm/CMakeCache.txt` contains:
```
//No help, variable specified on the command line.
CMAKE_MAKE_PROGRAM:UNINITIALIZED=MyNinjaB
```
and those are the commands being used for building with `cmake --build`

But of course, when running `ninja` directly instead of `cmake --build`, then `ninja` is being used.

Default value for  `CMAKE_MAKE_PROGRAM` is `ninja` when using `-GNinja`

Reply:
I see, but it's a moot point, it's too difficult to instruct Windows users to do this.

Reply:
> > cmake --build supports --, so you can do:
>
>I was not aware, could we have a secret MULTI_IMAGE_BUILD_OPT to populate it so I can debug then?

Let me do some digging

Reply:
> Perhaps we could have an overridable MULTI_IMAGE_BUILD_COMMMAND that defaulted to "cmake --build ." ?

We already have, as explained above, it's called: `-Dspm_CMAKE_MAKE_PROGRAM`

Reply:
Ignore that, it won't do.